### PR TITLE
Slugcat sentience

### PR DESCRIPTION
**Pull request by Walarks**
## About The Pull Request

Adds a component that turns a mob into a companion. Currently, this extends to fluff and is only enabled on the Newt. I could add gameplay-related stuff for it later, but that would require some more thought about balance™ and this stuff was ready so I thought of pushing it out instead of delaying it.

By using the xeno analyzer on the Newt, the Newt will now follow you and react to some voice commands.

To give a voice command, say `<mob-name>, <command>`. There is also a 10% chance for it to repeat anything you say when not saying its name.

The current commands are:
- `help` - the mob will pass you a note that contains the available commands
- `you will be called` - the mob will change its name to the next thing you say
- `repeat after me` - the mob will repeat the next thing you say
- `goodbye` - the mob will stop following you or listening to what you say

## Why It's Good For The Game

I've been wanting to see how having a mob following you and doing some stuff would play out. This shouldn't change anything combat-related, but can be something for people to interact with when there isn't much else to do.

## Changelog
:cl:
add: Adds a companion component and puts it on Newt, by using the analyzer on the Newt, it will react to some stuff you say. Say `Newt, help` and it will give you a list of commands.
/:cl:

### CODE REVIEW

**Comment by FosterGrime-Heymynameslouie** [comment]
RANGER THE SUPPORT DOG IS REAL

HE'S REALLLL NOWWW

File: code/datums/components/companion.dm [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+/datum/component/companion/proc/handle_master_change(datum/source, atom/movable/new_mob_master)
+	SIGNAL_HANDLER
+	if(mob_master == new_mob_master)
+		return
+
+	if(!new_mob_master)
+		unassign_mob_master("Farewell [mob_master]...")
+		return
+
+	var/master_update_message = "I shall follow you [new_mob_master]"
```
Comment (XSlayer300) [inline code/datums/components/companion.dm:69]:
```suggestion
	var/master_update_message = "I shall follow you [new_mob_master]."
```

**Comment by maikilangiolo** [comment]
Ok but make companions not bullet dense so they don't die when deployed

**Comment by Guidesu** [comment]
Cool! Now make them claimable by ghosts so people can finally be the ranger and carry items to marines with his little backpack and grab grenades with its mouth like ian on TG

**Comment by Losenis** [comment]
Newt, repeat after me, CURSE OF R[Redacted]

File: code/datums/components/companion.dm [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	if(mob_master)
+		unassign_mob_master()
+		master_update_message = "Farewell [mob_master]. I will now follow [new_mob_master]."
+
+	assign_mob_master(new_mob_master, master_update_message)
+
+///Handles assigning a new master mob
+/datum/component/companion/proc/assign_mob_master(atom/movable/new_mob_master, hello_message)
+	mob_master = new_mob_master
+	RegisterSignal(parent, COMSIG_MOVABLE_HEAR, .proc/handle_mob_master_speech)
```
Comment (TiviPlus) [inline code/datums/components/companion.dm:79]: missing registering in case the parent deletes

File: code/modules/mob/mob_helpers.dm [hunk 3]
```diff
@@ -471,3 +471,6 @@ mob/proc/get_standard_bodytemperature()
 	var/obj/item/inactive_item = get_inactive_held_item()
 	if(istype(inactive_item))
 		inactive_item.do_unique_action(src)
+
+/mob/proc/update_companion_master(mob/new_mob_master)
+	SEND_SIGNAL(src, COMSIG_MASTER_MOB_CHANGE, new_mob_master)
```
Comment (TiviPlus) [inline code/modules/mob/mob_helpers.dm:476]: couldnt you just replace this with registering for an attack signal, and then checking if the thing it's attacked by was a analyzer?

File: code/datums/components/companion.dm [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+/datum/component/companion/proc/goodbye()
+	unassign_mob_master("Farewell [mob_master]...")
+
+///Wrapper for say
+/datum/component/companion/proc/say(message)
+	SSspeech_controller.queue_say_for_mob(parent, message, SPEECH_CONTROLLER_QUEUE_SAY_VERB)
+
+///Wrapper for emote
+/datum/component/companion/proc/emote(message)
+	SSspeech_controller.queue_say_for_mob(parent, message, SPEECH_CONTROLLER_QUEUE_EMOTE_VERB)
```
Comment (TiviPlus) [inline code/datums/components/companion.dm:165]: Are these wrappers really needed

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Rest lgtm

File: code/datums/components/companion.dm (hunk 4 above)
Comment (Walarks) [reply code/datums/components/companion.dm:165]: I could make them defines

File: code/datums/components/companion.dm (hunk 2 above)
Comment (Walarks) [reply code/datums/components/companion.dm:80]: added

File: code/modules/mob/mob_helpers.dm (hunk 3 above)
Comment (Walarks) [reply code/modules/mob/mob_helpers.dm]: replaced with COMSIG_PARENT_ATTACKBY

**Review by TiviPlus — APPROVED** [review]

File: code/datums/components/companion.dm [hunk 5]
```diff
@@ -0,0 +1,169 @@
+///Wrapper for say
+#define SAY(message) SSspeech_controller.queue_say_for_mob(parent, message, SPEECH_CONTROLLER_QUEUE_SAY_VERB)
+///Wrapper for emote
+#define EMOTE(message) SSspeech_controller.queue_say_for_mob(parent, message, SPEECH_CONTROLLER_QUEUE_EMOTE_VERB)
```
[the remark this answers is not in the document]
Comment (Walarks) [reply code/datums/components/companion.dm]: updated

**Review by Lumipharon — APPROVED** [review]

**Pull request merged**